File: src/24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
基本思想-多指针
吐了，做成逆置单链表了，诶
修改，可以递归的做
time defeat: 76%
space defeat: 75%
time consuming: 1 hour
'''
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        #空链表或者只有一个节点 
        if not head or not head.next:
            return head
        #否则，原先的第二个节点就是新链表的头
        newHead = head.next
        #下一个节点就是缩小递归范围后的第三个节点开始的链表
        head.next = self.swapPairs(newHead.next)
        #将本来的头加到第二个节点
        newHead.next = head
        return newHead
        '''
            #大于等于三个节点的情况
            lastPreNode = None
            firstNode = head
            prePointer = head.next
            sufPointer = head.next.next
            while sufPointer != None or prePointer != None:
                print(firstNode.val, prePointer.val)
                #逆置相邻两个节点
                prePointer.next = firstNode
                firstNode.next = lastPreNode
                #更新指针
                lastPreNode = prePointer
                firstNode = sufPointer
                if not sufPointer:
                    break
                prePointer = sufPointer.next
                sufPointer = prePointer.next
            return prePointer
        '''
        
        

File: src/24.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        //好久之前做的了，干翻腾讯CSIG！
        //要保存prev和next
        //当prev->next为空时完蛋
        //用个假头节点就很方便
        //很棒！一次提交，二十五分钟
        //做的完全正确，如果不是交换节点值，而是更改链接，接近一个困难题了
        //自己还是做出来更改链接的版本了
        //边界判断
        if(head == nullptr || head->next == nullptr){
            return head;
        }
        //假头节点出来
        ListNode * fakeHead = new ListNode(0);
        ListNode * prev = fakeHead;
        prev->next = head;
        //要求每段要有两个子节点
        while(prev->next && prev->next->next){
            ListNode * node1 = prev->next;
            ListNode * node2 = node1->next;
            ListNode * next = node2->next;
            //更换链接
            prev->next = node2;
            node2->next = node1;
            node1->next = next;
            //前进
            prev = node1;
            //现在考虑奇数个节点的情况
        }
        return fakeHead->next;
    }   
};
